### PR TITLE
Change UfsManager key to not rely on properties

### PR DESCRIPTION
**Pull request by jiacheliu3**
### What changes are proposed in this pull request?

In the UfsManager the UFS properties are not a prerequisite for locating an UFS instance. By not hashing on properties, the `get()` does not need to resolve the target UFS property keys, which saves cost.

Before: 15.32% CPU in a worker stress test
image

After: <1% CPU in the same worker stress test
image

### CODE REVIEW

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java [hunk 1]
```diff
@@ -91,13 +91,11 @@ public DoraMetaManager(AlluxioConfiguration conf,
     mUfsManager = ufsManager;
   }
 
-  private UnderFileSystem getUfsInstance(String ufsUriStr) {
+  protected UnderFileSystem getUfsInstance(String ufsUriStr) {
     AlluxioURI ufsUriUri = new AlluxioURI(ufsUriStr);
     try {
       UnderFileSystem ufs = mUfsManager.getOrAdd(ufsUriUri,
-          // todo(bowen): local configuration may not have UFS-specific configurations
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/DoraMetaManager.java:98]: this todo is no longer relevant

File: dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java [hunk 2]
```diff
@@ -207,13 +207,10 @@ protected PagedDoraWorker(
   }
 
   @VisibleForTesting
-  protected UnderFileSystem getUfsInstance(String ufsUriStr) {
+  public UnderFileSystem getUfsInstance(String ufsUriStr) {
```
Comment (jiacheliu3) [inline dora/core/server/worker/src/main/java/alluxio/worker/dora/PagedDoraWorker.java:210]: allow objects holding the worker ref to getUfsInstance. This helps our next step dep injection refactors.

File: dora/core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 3]
```diff
@@ -230,11 +231,10 @@ public void removeMount(long mountId) {
    * Gets an instance for the given UFS URI and configuration, if such exists.
    *
    * @param ufsUri the URI of the UFS
-   * @param ufsConf the configuration for the UFS
    * @return a UFS instance, or none if not registered
    */
-  public Optional<UnderFileSystem> get(AlluxioURI ufsUri, UnderFileSystemConfiguration ufsConf) {
-    Key key = generateKey(ufsUri, ufsConf);
+  public Optional<UnderFileSystem> get(AlluxioURI ufsUri) {
+    Key key = generateKey(ufsUri);
```
Comment (jja725) [inline dora/core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java:237]: The old mount table may still need properties...

**Review by jja725 — COMMENTED** [review]
Let's wait for a while until we figure out whether we really don't need properties

File: dora/core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java [hunk 4]
```diff
@@ -116,44 +113,48 @@ protected AbstractUfsManager() {
...
    *
    * @param ufsUri  the UFS path
-   * @param ufsConf the UFS configuration
+   * @param ufsConfSupplier supplier for UFS configuration
    * @return the UFS instance
    */
-  public UnderFileSystem getOrAdd(AlluxioURI ufsUri, UnderFileSystemConfiguration ufsConf) {
-    Key key = generateKey(ufsUri, ufsConf);
+  public UnderFileSystem getOrAdd(AlluxioURI ufsUri,
+                                  Supplier<UnderFileSystemConfiguration> ufsConfSupplier) {
```
Comment (yyongycy) [inline dora/core/server/common/src/main/java/alluxio/underfs/AbstractUfsManager.java:120]: ident ?

**Comment by jiacheliu3** [comment]
I further checked 
This change does not change how a mount point is added -- you still add a mount point with <mount point path, mount options>. The change is how you find the corresponding UFS instance.

Before this change: 
You can have multiple UFS instances in a map like this:
```
Map:
Key<Ufs path, Ufs options> -> UFS instance
Key<Ufs path, Ufs options> -> UFS instance
Key<Ufs path, Ufs options> -> UFS instance
```
Theoretically you may have multiple UFS instances on the same UFS path, but having different UFS options. But later in this section I will explain why no code is doing that.

After this change:
You can have multiple UFS instances in a map like this:
```
Map:
Key<Ufs path> -> UFS instance
Key<Ufs path> -> UFS instance
Key<Ufs path> -> UFS instance
```
So you see there's no way you can use multiple UFS instances on the same path but different UFS options.

I checked in the code, two code paths can add to this map:
1. `AbstractUfsManager.addMount()/addMountWithRecorder()`
2. `AbstractUfsManager.getOrAdd()`

For `1`, that means the values are only added from mount operations. It is either `OldMountCommand` or in Dora the `alluxio.dora.client.ufs.root`. Either way, before a new UFS is put into the map with a key, we verify that the mount path does NOT already exist. All usages of `addMount()` are shown in the screenshot below:
Screen Shot 2023-10-14 at 7 52 12 PM

For `2`, all usages are in the Dora worker. In a Dora worker (in OS), even when we create the `Key(Ufs path, Ufs options)`, the ufs options are only `Configurations.global()`.
Screen Shot 2023-10-14 at 6 42 57 PM

To summarize, we are not losing any functionalities here by this change.

**Review by jja725 — APPROVED** [review]
LGTM

**Comment by jiacheliu3** [comment]
Out of curiosity, I validated the identity function before this change -- `Key(AlluxioURI, UnderFileSystemConfiguration)`. It works, by:
1. If two objects have different AlluxioURI, the identities are different.
2. If two objects have the same AlluxioURI and config properties contain different kv, the identities are different.
3. If two objects have the same AlluxioURI, and the config properties contain same kvs, although the config properties are in different HashMaps and the maps are not ordered, the identities are equal.

Although 3 looks very reasonable, the implementation was not so clear and straightforward. For `HashMap`, the `equals()` method is from `AbstractMap.equals()`, which compares entries one by one by checking `contains()`. That's the root reason why two different maps with different orders can equal to one another.

I'm leaving this comment just as a trace, so we know what identity function might work if we later decide to add it back.

**Review by Jackson-Wang-7 — APPROVED** [review]
LGTM

**Comment by jiacheliu3** [comment]
alluxio-bot, merge this please

**Comment by yyongycy** [comment]
@yyongycy  Mark here. We might need this in some corner case.
